verifier/verifier.go

```diff
@@ -0,0 +1,97 @@
+package verifier
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/openpubkey/openpubkey/pktoken"
+)
+
+type Verifier interface {
+	AddCosignerVerifier(issuer string, strict bool)
+	VerifyPKToken(ctx context.Context, pkt *pktoken.PKToken, checks ...Check) error
+}
+
+type Check func(*pktoken.PKToken) error
+
+type cosignerOpts struct {
+	issuer string
+	strict bool
+}
+
+type verifier struct {
+	issuer           string
+	commitmentClaim  string
+	cosignerVerifier cosignerOpts
+}
+
+func New(issuer, commitmentClaim string) Verifier {
+	return &verifier{
+		issuer:          issuer,
```

Comment:
> Your motivation here is to not depend on client.provider because the verifier package should be using structs from the client package for reasons of architecture and cyclical dependences?

The Verifier can't use the same interface as the Client because they require different things. Defining an interface for the Verifier would probably look like
```golang
type ProviderVerifier interface {
    Issuer() string
    CommitmentClaim() string
}
```
which is just an unnecessary complex wrapper around two constants.

> New needs more than the issuer to verify the ID Token, we need to know the aud/client-id as well to check it on the ID Token.

> Maybe it makes sense to have some sort of verifier struct/interface that has issuer, audience, and a JWKS resolver function. The JWKS resolver function could then be extended for looking up archival OP pubkeys.

These are good points that I hadn't thought about including, but make sense to have as options.

Reply:
> The Verifier can't use the same interface as the Client because they require different things

What about having a common OP struct that gets different parts used by two different interfaces: a client interface and a verifier interface. The advantage of that approach is someone adding a new provider only has to create one file and all the information about that provider is stored together.

> which is just an unnecessary complex wrapper around two constants and a bit awkward to make configurable.

With Google you need at least three constants: the issuer, the audience and the commitment claim. If we added gitlab we would need the ability to tell it to not have a commitment claim and add the CIC via the GQ signature. So I guess what I'm saying here is that it would be nice to have a per OP function which is called that makes this decisions in a flexible and programmable way.

> These are good points that I hadn't thought about including, but make sense to have as options.

Agreed and I believe that was the motivation behind having the PublicKey() method for OpenID Providers. You can use a decorator or just implement a new provider on the interface to point it  to an archival JWKS.